File: app/src/main/java/igrek/todotree/services/commander/SecretCommander.java
package igrek.todotree.services.commander;


import java.util.Arrays;
import java.util.List;

import igrek.todotree.logger.Logs;
import igrek.todotree.services.preferences.Preferences;
import igrek.todotree.services.preferences.PropertyDefinition;
import igrek.todotree.services.resources.UserInfoService;

/**
 * secret commands parser:
 * ###dupa
 * ###config lockdb true
 * ###config lockdb false
 * ###config dbFilePath /storage/6D49-845B/Android/data/igrek.todotree/todo.json
 */
public class SecretCommander {
	
	private Logs logger;
	private Preferences preferences;
	private UserInfoService userInfo;
	
	private static final String CMD_PREFIX = "###";
	
	public SecretCommander(Logs logger, Preferences preferences, UserInfoService userInfo) {
		this.logger = logger;
		this.preferences = preferences;
		this.userInfo = userInfo;
	}
	
	public boolean execute(String content) {
		if (!content.startsWith(CMD_PREFIX))
			return false;
		
		String commandsAll = content.substring(CMD_PREFIX.length());
		if (commandsAll.isEmpty())
			return false;
		
		return executeCommand(Arrays.asList(commandsAll.split(" ")));
	}
	
	private boolean executeCommand(List<String> parts) {
		String main = parts.get(0);
		if (main.isEmpty())
			return false;
		List<String> params = parts.subList(1, parts.size());
		
		switch (main) {
			case "dupa":
				commandDupa(params);
				break;
			case "config":
				commandConfig(params);
				break;
			default:
				userInfo.showInfo("Unknown command: " + main);
				return false;
		}
		return true;
	}
	
	private void commandDupa(List<String> params) {
		userInfo.showInfo("Congratulations! You have discovered an Easter Egg.");
	}
	
	private void commandConfig(List<String> params) {
		if (params.size() < 2) {
			logger.warn("not enough params");
			return;
		}
		String property = params.get(0).toLowerCase();
		String value = params.get(1);
		if (property.equals("lockdb")) {
			value = value.toLowerCase();
			boolean valueB = value.equals("true") || value.equals("on") || value.equals("1");
			preferences.setValue(PropertyDefinition.lockDB, valueB);
			userInfo.showInfo("Property saved: lockdb = " + valueB);
		} else if (property.equals("dbfilepath")) {
			preferences.setValue(PropertyDefinition.dbFilePath, value);
			userInfo.showInfo("Property saved: dbFilePath = " + value);
		} else {
			logger.warn("unknown property: " + property);
		}
		
	}
	
}

File: app/src/main/java/igrek/todotree/service/commander/SecretCommander.kt
package igrek.todotree.service.commander

import igrek.todotree.info.UiInfoService
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import igrek.todotree.intent.ItemEditorCommand
import igrek.todotree.settings.SettingsState
import java.util.*

/**
 * secret commands parser:
 * ###dupa
 * ###config lockdb true
 * ###config lockdb false
 * ###config userauthtoken authToken
 * ###remote_item
 */
class SecretCommander(
    settingsState: LazyInject<SettingsState> = appFactory.settingsState,
    uiInfoService: LazyInject<UiInfoService> = appFactory.uiInfoService,
) {
    private val settingsState by LazyExtractor(settingsState)
    private val uiInfoService by LazyExtractor(uiInfoService)

    private val logger = LoggerFactory.logger

    fun execute(content: String): Boolean {
        if (!content.startsWith(CMD_PREFIX)) return false
        val commandsAll = content.substring(CMD_PREFIX.length)
        return if (commandsAll.isEmpty()) false else executeCommand(Arrays.asList(*commandsAll.split(" ").toTypedArray()))
    }

    private fun executeCommand(parts: List<String>): Boolean {
        val main = parts[0]
        if (main.isEmpty()) return false
        val params = parts.subList(1, parts.size)
        when (main) {
            "dupa" -> commandDupa(params)
            "config" -> commandConfig(params)
            "remote_item" -> createRemoteItem()
            else -> {
                uiInfoService.showSnackbar("Unknown command: $main")
                return false
            }
        }
        return true
    }

    private fun createRemoteItem() {
        ItemEditorCommand().createRemoteItem()
    }

    private fun commandDupa(params: List<String>) {
        uiInfoService.showSnackbar("Congratulations! You have discovered an Easter Egg.")
    }

    private fun commandConfig(params: List<String>) {
        if (params.size < 2) {
            logger.warn("not enough params")
            return
        }
        val property = params[0].toLowerCase()
        var value = params[1]
        when (property) {
            "lockdb" -> {
                value = value.toLowerCase()
                val valueB = value == "true" || value == "on" || value == "1"
                settingsState.lockDB = valueB
                uiInfoService.showSnackbar("Property saved: $property = $valueB")
            }
            "userauthtoken" -> {
                settingsState.userAuthToken = value
                uiInfoService.showSnackbar("Property saved: $property = $value")
            }
            else -> {
                logger.warn("unknown property: $property")
            }
        }
    }

    companion object {
        private const val CMD_PREFIX = "###"
    }

}